lib/kube/proxy/pod_filters.go

```diff
@@ -0,0 +1,325 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/kube/proxy/responsewriters"
+)
+
+// newPodFiltererBuilder creates a wrapper function that once executed creates
+// a runtime filter for Pods.
+// The filter exclusion criteria is:
+// - deniedPods: excluded if (namespace,name) matches an entry even if it matches
+// the allowedPod's list.
+// - allowedPods: excluded if (namespace,name) not match a single entry.
+func newPodFiltererBuilder(allowedPods, deniedPods []types.KubernetesResource, log logrus.FieldLogger) responsewriters.FilterWrapper {
+	negotiator := newClientNegotiator()
+	return func(contentType string, responseCode int) (responsewriters.Filter, error) {
+		return newPodFilterer(
+			podFiltererConfig{
+				contentType:      contentType,
+				responseCode:     responseCode,
+				negotiator:       negotiator,
+				allowedResources: allowedPods,
+				deniedResources:  deniedPods,
+				log:              log,
+			},
+		)
+	}
+}
+
+// podFiltererConfig holds the configuration for filtering pods from responses.
+type podFiltererConfig struct {
+	// contentType is the response "Content-Type" header.
+	contentType string
+	// responseCode is the response status code.
+	responseCode int
+	// negotiator is an instance of a client negotiator.
+	negotiator runtime.ClientNegotiator
+	// allowedResources is the list of kubernetes resources the user has access to.
+	allowedResources []types.KubernetesResource
+	// deniedResources is the list of kubernetes resources the user must not access.
+	deniedResources []types.KubernetesResource
+	// log is the logger.
+	log logrus.FieldLogger
+}
+
+// newPodFilterer creates a new instance of podFilterer with the encoders and
+// decoders for the given contentType value.
+// podFilterer is able to exclude Pods from list requests executed against the
+// target cluster.
+// The filter exclusion criteria is:
+// - deniedResources: excluded if (namespace,name) matches an entry even if it matches
+// the allowedResources's list.
+// - allowedResources: excluded if (namespace,name) not match a single entry.
+func newPodFilterer(cfg podFiltererConfig) (*podFilterer, error) {
+	encoder, decoder, err := newEncoderAndDecoderForContentType(cfg.contentType, cfg.negotiator)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &podFilterer{
+		encoder:           encoder,
+		decoder:           decoder,
+		podFiltererConfig: cfg,
+	}, nil
+}
```

Comment:
Simplify?

I think we could do away with podFiltererConfig too, seems unnecessary.

```suggestion
// newPodFilterer creates a wrapper function that once executed creates
// a runtime filter for Pods.
// The filter exclusion criteria is:
// - deniedPods: excluded if (namespace,name) matches an entry even if it matches
// the allowedPod's list.
// - allowedPods: excluded if (namespace,name) not match a single entry.
func newPodFilterer(allowedPods, deniedPods []types.KubernetesResource, log logrus.FieldLogger) responsewriters.FilterWrapper {
	return func(contentType string, responseCode int) (responsewriters.Filter, error) {
		cfg := podFiltererConfig{
			contentType:      contentType,
			responseCode:     responseCode,
			negotiator:       newClientNegotiator(),
			allowedResources: allowedPods,
			deniedResources:  deniedPods,
			log:              log,
		}

		encoder, decoder, err := newEncoderAndDecoderForContentType(cfg.contentType, cfg.negotiator)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		return &podFilterer{
			encoder:           encoder,
			decoder:           decoder,
			podFiltererConfig: cfg,
		}, nil
	}
}

// podFiltererConfig holds the configuration for filtering pods from responses.
type podFiltererConfig struct {
	// contentType is the response "Content-Type" header.
	contentType string
	// responseCode is the response status code.
	responseCode int
	// negotiator is an instance of a client negotiator.
	negotiator runtime.ClientNegotiator
	// allowedResources is the list of kubernetes resources the user has access to.
	allowedResources []types.KubernetesResource
	// deniedResources is the list of kubernetes resources the user must not access.
	deniedResources []types.KubernetesResource
	// log is the logger.
	log logrus.FieldLogger
}

// podFilterer is a pod filterer instance.
type podFilterer struct {
	podFiltererConfig
	encoder runtime.Encoder
	decoder runtime.Decoder
}
```